install/gcp/installer/util/gcloud.py

```diff
@@ -791,7 +790,7 @@ def check_vm_init_status(vm_name, zone):
         bool: Whether or not the VM has finished initializing.
     """
 
-    check_script_executed = 'tail -n1 /tmp/deployment.log'
+    check_script_executed = '\"tail -n1 /tmp/deployment.log\"'
```

Comment:
If I am not mistaken, this is working on cloudshell. Can you verify by adding the quotes it doesn't break on cloudshell? Also, why is the escape required in this case i.e. '\"tail -n1 /tmp/deployment.log\"' instead of '"tail -n1 /tmp/deployment.log"'?

Reply:
The unescaped version was throwing linter errors locally, so I escaped them. If it'll pass your linting, I'm more than happy to change that :).

Output verifying behavior in Cloudshell:

john@johnevans-forseti-test2:~$ gcloud compute ssh forseti-server-vm-8452c81 --zone us-central1-c --command tail -n1 /tmp/deployment.log --quiet
ERROR: (gcloud.compute.ssh) unrecognized arguments:
  -n1
  /tmp/deployment.log

vs

gcloud compute ssh forseti-server-vm-8452c81 --zone us-central1-c --command "tail -n1 /tmp/deployment.log" --quiet
WARNING: The public SSH key file for gcloud does not exist.
WARNING: The private SSH key file for gcloud does not exist.
WARNING: You do not have an SSH key for gcloud.
WARNING: SSH keygen will be executed to generate a key.
Generating public/private rsa key pair.
Your identification has been saved in /home/john/.ssh/google_compute_engine.
Your public key has been saved in /home/john/.ssh/google_compute_engine.pub.
The key fingerprint is:
The key's randomart image is:
+---[RSA 2048]----+
+----[SHA256]-----+
Updating project ssh metadata...⠶Updated [https://www.googleapis.com/compute/v1/projects/johnevans-forseti-test2].
Updating project ssh metadata...done.
Waiting for SSH key to propagate.
Warning: Permanently added 'compute.8919612972468928729' (ECDSA) to the list of known hosts.
Execution of startup script finished

I wonder if maybe what happened here is a change instrinsic to gcloud. I am running 209.0.0 and I believe Morgante is running the same.

Reply:
Hmmm interesting. I just reran with the explicitly escaped quotes removed and it fails with:

tail -n1 /tmp/deployment.log: No such file or directory

and the server never registers as being up (even though it's clearly up and running). Adding the escaped quotes back doesn't seem to make a difference. So now I'm not sure. Removing the quotes entirely, and the server comes up, but I start seeing errors in the logs again... 

I need to rethink this.

Reply:
Hi @johnrevans6 ,

I understand you won't be able to run the command directly off cloudshell, but inside the installer, it is running through the run_command method defined [here]{https://github.com/GoogleCloudPlatform/forseti-security/blob/ba1e3d4a10c0c6076012963b581fcde46242b792/install/gcp/installer/util/utils.py#L357} and it should treat that argument as one string.

I just tried running the installer on cloudshell again and everything seems to be working, can you please clarify?

Thanks,
Joe

Reply:
I agree. I wonder if it had something to do with the version of gcloud we were using. After I updated, I stopped seeing errors in the logs. Waiting for Morgante to confirm on his end.

Reply:
Thanks for the update!

Reply:
I updated gcloud to the latest version and still see this - it's not clear why this would cause any regression in old versions though? I'd assume that including the entire command as a single arg is less error-prone than hoping gcloud correctly handles multiple arguments (it's not even clear to me how it's meant to know those are all arguments to --commmand).

@joecheuk I see that this is run through run_command, but even though Python sees this as a single string the ultimate output will still be `gcloud compute ssh forseti-server-vm-8452c81 --zone us-central1-c --command tail -n1 /tmp/deployment.log --quiet`. Honestly, I'm surprised this would ever work—how does gcloud know the "--quiet" argument is for it instead of for tail?

Reply:
Hi @morgante ,

I just verified again on cloud shell by executing the command directly using python's subprocess module and everything seems to be working.

>>> import subprocess
>>> a = ['gcloud', 'compute', 'ssh', 'forseti-server-vm-290bd72', '--zone', 'us-central1-c', '--command', 'tail -n1 /tmp/deployment.log --quiet']
>>>
>>>
>>> subprocess.Popen(a)
<subprocess.Popen object at 0x7f0835b74210>
>>> Execution of startup script finished

Would you mind providing the steps to reproduce the problem you are seeing?

Reply:
@joecheuk I just updated my gcloud version and it looks like this is working fine. Sorry for the confusion.

Reply:
No problem at all, thanks for the update!

Reply:
Cool. What I'll go ahead and change this back, then. Thanks everybody for their feedback.